uamqp/aio/_session_async.py

```diff
@@ -204,24 +209,75 @@ async def _incoming_flow(self, frame):
         if frame[4] is not None:
             await self._input_handles[frame[4]]._incoming_flow(frame)
         else:
-            async with create_task_group() as tg:
-                for link in self._output_handles.values():
-                    if self.remote_incoming_window > 0 and not link._is_closed:
-                        await tg.spawn(link._incoming_flow, frame)
+            futures = []
+            for link in self._output_handles.values():
+                if self.remote_incoming_window > 0 and not link._is_closed:
+                    futures.append(link._incoming_flow(frame))
+            await asyncio.gather(*futures)
```

Comment:
On another note - we had talked about making Connection/Session/Link objects threadsafe - this will also apply to concurrency-safe.
Wherever we are using `aynscio.gather` - each of those Sessions/Links are behaving concurrently - which could lead to race conditions if they do anything to modify their parent (e.g. if one link errors and updates the session state, a concurrent link might try to set a conflicting session state). 
So we may need to revisit those discussions in this context.

Reply:
yes, thanks for reminding me of the race condition! this is something that we should definitely address..

since currently we are not sharing session/connection, I think we could log a separate issue for race condition and address it in a future PR?

Reply:
issue created: https://github.com/Azure/azure-sdk-for-python/issues/22973